keps/sig-autoscaling/20190307-configurable-scale-velocity-for-hpa.md

```diff
@@ -0,0 +1,446 @@
+[Algorithm details]: https://v1-14.docs.kubernetes.io/docs/tasks/run-application/horizontal-pod-autoscale/#algorithm-details
+
+#### Story 2: Scale Up As Fast As Possible, Scale Down Very Gradually
+
+This mode is essential when you don’t want to risk scaling down very quickly.
+
+Create an HPA with the constraints:
+
+- `constraints`:
+  - `scaleUp`:
+    - `percent = 900` (i.e. increase number of pods 10 times per minute is ok).
+  - `scaleDown`:
+    - `pods = 1`
+    - `periodSeconds = 600` (i.e., scale down one pod every 10 min)
+
+All other parameters are not specified (default values are used)
+
+If the application is started with 1 pod, it will scale up with the following number of pods:
+
+    1 -> 10 -> 100 -> 1000
+
+So, it can reach `maxReplicas` very fast.
+
+Scaling down will be by one pod each 10 min:
+
+    1000 -> 1000 -> 1000 -> … (7 more min) -> 999
+
+#### Story 3: Scale Up Very Gradually, Usual Scale Down Process
+
+This mode is essential when you want to increase capacity, but you want it to be very pessimistic.
+
+Create an HPA with the constraints:
+
+- `constraints`:
+  - `scaleUp`:
+    - `pods = 1`    (i.e., increase only by one pod per minute)
+
+All other parameters are not specified (default values are used)
+
+If the application is started with 1 pod, it will scale up very gradually:
+
+    1 -> 2 -> 3 -> 4
+
+Scale down will be done a usual way (check stabilization window in [Algorithm details][])
+
+[Algorithm details]: https://v1-14.docs.kubernetes.io/docs/tasks/run-application/horizontal-pod-autoscale/#algorithm-details
+
+#### Story 4: Scale Up As Usual, Do Not Scale Down
+
+This mode is essential when you don’t want to risk scaling down at all.
+
+Create an HPA with the following constraints:
+
+- `constraints`:
+  - `scaleDown`:
+    - `percent = 0`
+    - `pods = 0`
+
+i.e., set both constraints to 0, so that the HPA controller would never scale the cluster down
+
+All other parameters are not specified (default values are used)
+
+The cluster will scale up as usual (default values), but will never scale down.
+
+#### Story 5: Delay Before Scaling Down
+
+This mode is used when the user expects a lot of flapping
+or doesn't want to turn off pods too early expecting some late load spikes.
+
+Create an HPA with the following constraints:
+
+- `constraints`:
+  - `scaleDown`:
+    - `pods = 5`
+  - `delaySeconds = 600`
+
+i.e., the algorithm will:
+
+- gather recommendations for 600 seconds (by default: 300)
+- pick the largest one
+- turn off no more than 5 pods per minute
+
+### Implementation Details/Notes/Constraints
+
+#### Algorithm Pseudocode
+
+The algorithm to find the number of pods will look like this:
+
+```golang
+desiredReplicas = 0
+calculatedReplicas = AnyAlgorithmInHPAController(...)
+if calculatedReplicas > curReplicas:
+  constraint1 = CurReplicas * (1 + ScaleUpPercent/100)
+  constraint2 = CurReplicas + ScaleUpPods
+  scaleUpLimit = max(constraint1, constraint2)
```

Comment:
There is one subtle change required here.  It makes sense to use the `max` of rate limits when scaling up to allow the autoscaler to get off the ground.  But when combining rate with delay, you should use the minimum.

Reply:
(chatting with @gliush about this)

Ivan already had this in mind.  In the delay pseudo code below, it already uses the min to make the delay effective.  So you can disregard this comment.

Reply:
:+1: